File: python/src/loopiter/testing.py
"""Executable adapter conformance checks. Use only an explicitly disposable namespace."""

import asyncio
from copy import deepcopy
from uuid import uuid4

from . import FeedbackLoop, LoopiterError
from .store import FeedbackStore


async def run_store_conformance(store: FeedbackStore, *, namespace: str | None = None) -> None:
    """Asserts transaction, revision, idempotency, isolation and pagination guarantees.

    Creates/deletes its own unique namespaces. Supplying a namespace is an explicit
    request to clear that namespace; never pass a production namespace.
    Does not close the developer-owned store/pool. Raises AssertionError on failure.
    """
    namespace = namespace or f"loopiter-conformance/{uuid4()}"
    other = namespace + "/isolated"
    loop = FeedbackLoop(store=store, namespace=namespace)
    isolated = FeedbackLoop(store=store, namespace=other)
    try:
        first = await loop.record_execution(id="a", kind="prediction")
        assert await loop.record_execution(id="a", kind="prediction") == first
        assert await isolated.get_execution("a") is None
        await isolated.record_execution(id="a", kind="tool")
        try:
            await loop.record_execution(id="a", kind="tool")
        except LoopiterError as e:
            assert e.code == "conflict"
        else:
            raise AssertionError("Conflicting insert overwrote an existing record")
        results = await asyncio.gather(
            *(
                loop.complete_execution("a", expected_revision=1, output={"attempt": i})
                for i in range(2)
            ),
            return_exceptions=True,
        )
        assert sum(isinstance(r, dict) for r in results) == 1
        assert sum(isinstance(r, LoopiterError) and r.code == "conflict" for r in results) == 1
        current = await loop.get_execution("a")
        async with store.transaction(namespace) as tx:
            leaked = tx
            record = deepcopy(current)
            record.update(id="b", revision=1)
            await tx.insert("executions", record)
        try:
            await leaked.get("executions", "a")
        except LoopiterError as e:
            assert e.code == "transaction_closed"
        else:
            raise AssertionError("Completed transaction remained usable")
        try:
            async with store.transaction(namespace) as tx:
                record = deepcopy(first)
                record["id"] = "rolled-back"
                await tx.insert("executions", record)
                raise RuntimeError("injected rollback")
        except RuntimeError:
            pass
        assert await loop.get_execution("rolled-back") is None
        async with store.transaction(namespace) as tx:
            returned = await tx.get("executions", "a")
            returned["kind"] = "custom"
        assert (await loop.get_execution("a"))["kind"] == "prediction"
        page = await loop.list("executions", limit=1)
        assert [r["id"] for r in page["items"]] == ["a"]
        assert [
            r["id"] for r in (await loop.list("executions", cursor=page["next_cursor"]))["items"]
        ] == ["b"]
        signal = await loop.record_signal(
            id="signal", execution_id="a", kind="rating", name="correct", value=True, source="test"
        )
        try:
            async with store.transaction(namespace) as tx:
                signal["revision"] += 1
                await tx.replace("signals", signal, 1)
        except LoopiterError as e:
            assert e.code == "immutable_record"
        else:
            raise AssertionError("Signal was mutable")
        try:
            async with store.transaction(other) as tx:
                await tx.insert("executions", first)
        except LoopiterError as e:
            assert e.code == "namespace_mismatch"
        else:
            raise AssertionError("Cross-namespace write accepted")
        await loop.delete_namespace(confirmation=namespace)
        assert await loop.get_execution("a") is None
        assert (await isolated.get_execution("a"))["kind"] == "tool"
    finally:
        await store.delete_namespace(namespace)
        await store.delete_namespace(other)

File: src/testing.ts
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";
import type { FeedbackStore, StoreTransaction } from "./store.js";
import { FeedbackLoop } from "./feedback-loop.js";
/** Reusable, destructive ONLY to random conformance namespaces it creates. */
export async function runStoreConformance(
  store: FeedbackStore,
): Promise<string[]> {
  const namespace = `conformance/${randomUUID()}`,
    other = `${namespace}/other`;
  const loop = new FeedbackLoop({ store, namespace }),
    second = new FeedbackLoop({ store, namespace: other });
  const results: string[] = [];
  try {
    const record = await loop.recordExecution({ id: "a", kind: "turn" });
    await second.recordExecution({ id: "a", kind: "prediction" });
    assert.equal((await loop.getExecution("a"))?.kind, "turn");
    results.push("namespace isolation");
    assert.equal(
      (await loop.recordExecution({ id: "a", kind: "turn" })).revision,
      1,
    );
    await assert.rejects(loop.recordExecution({ id: "a", kind: "tool" }));
    results.push("idempotency and conflicts");
    await assert.rejects(
      store.transaction(namespace, async (tx) => {
        await tx.insert("executions", { ...record, id: "rolled-back" });
        throw new Error("abort");
      }),
    );
    assert.equal(await loop.getExecution("rolled-back"), undefined);
    results.push("transaction rollback");
    const updates = await Promise.allSettled([
      loop.completeExecution("a", { output: 1 }, 1),
      loop.completeExecution("a", { output: 2 }, 1),
    ]);
    assert.equal(updates.filter((r) => r.status === "fulfilled").length, 1);
    results.push("revision concurrency");
    await loop.recordExecution({ id: "b", kind: "turn" });
    const page = await loop.list("executions", { limit: 1 });
    assert.equal(page.items.length, 1);
    assert.ok(page.nextCursor);
    const next = await loop.list("executions", {
      limit: 1,
      cursor: page.nextCursor,
    });
    assert.equal(next.items[0]?.id, "b");
    results.push("pagination");
    let escaped: StoreTransaction | undefined;
    await store.transaction(namespace, async (tx) => {
      escaped = tx;
    });
    await assert.rejects(escaped!.get("executions", "a"));
    results.push("transaction lifetime");
    await assert.rejects(
      store.transaction(namespace, (tx) =>
        tx.insert("executions", { ...record, namespace: other, id: "cross" }),
      ),
    );
    results.push("write isolation");
    for(const kind of ['runs','operations','budgets','observations','coordination'] as const){
      const row={id:'controller',namespace,revision:1,createdAt:loop.now(),updatedAt:loop.now(),format:1 as const,data:{value:1,modelRequests:0,tokens:0,deployments:0}};
      await store.transaction(namespace,tx=>tx.insert(kind,row));
      assert.equal((await loop.list(kind)).items.length,1);
      assert.equal((await second.list(kind)).items.length,0);
      if(kind==='observations')await assert.rejects(store.transaction(namespace,tx=>tx.replace(kind,{...row,revision:2},1)));
      else {
        const updates=await Promise.allSettled([1,2].map(value=>store.transaction(namespace,tx=>tx.replace(kind,{...row,revision:2,data:{...row.data,value}},1))));
        assert.equal(updates.filter(r=>r.status==='fulfilled').length,1);
      }
    }
    results.push('autonomy collections, revisions and immutable observations');
    await loop.deleteNamespace(namespace);
    assert.equal(await loop.getExecution("a"), undefined);
    assert.ok(await second.getExecution("a"));
    results.push("namespace deletion");
    return results;
  } finally {
    await store.deleteNamespace(namespace);
    await store.deleteNamespace(other);
  }
}
